contrib/utilities/indent

```diff
@@ -21,6 +21,20 @@
 ##
 ## ---------------------------------------------------------------------
 
+find_astyle() {
+  IFS=':' read -r -a PATHS <<< "$PATH"
+
+  astyle=""
+  while IFS= read -r -d '' exe; do
+    astyle="${exe}"
+    if test "$("${astyle}" --version 2>&1)" == "Artistic Style Version 2.04" ; then
+      break
+    fi
+  done < <(find "${PATHS[@]}" -iname '*astyle*' -print0 2>/dev/null)
```

Comment:
Interesting construct. I think it would be easier to read if you did
```
  find ... | (while ... ; do ... ; done)
```

Reply:
Or is that not the same? I think the style most people would use is
```
  ASTYLE_PATHS=`find ...`
  for i in $ASTYLE_PATHS ; do ... ; done
```

Reply:
There's a bunch of ways to do this. The first approach `(find... | while ...)` works, but if I remember correctly it has some weird scope thing about variables inside the while loop. I might be wrong about that. The second one `(ASTYLE_PATHS=...; for i in $ASTYLE_PATHS...)` is dangerous if there are spaces in any of the path names. That can be fixed with arrays, and careful use of quote marks I think.

On the one hand, only a maniac would put spaces in their executable binary file names. But on the other hand, if you have a bunch of non-standard PATH entries I could see it happening.

tl;dr: Whatever you think is most clear. Let me know what you prefer.

Reply:
Fair points. I'm just not (yet!) familiar with the `while ... done < <(find ...)` construct :-) I learn something new every day!

@tjhei @gassmoeller -- any thoughts?

Reply:
I have to say I do not understand the `while ... done < <` construct (and I would guess most readers of the file wouldnt, on the other hand there will not be many readers of such a file). I am just not very much into bash scripting, I know how to do simple loops, pipes and functions. Does this pipe the results of the `find` into the `done` command, which magically means it is available inside the loop? And why the two separate `<` marks? If I had written it I would go with `ASTYLE_PATHS=; for i in $ASTYLE_PATHS` and use an array or enclose each argument in quotations (dont know exactly how complicated that would be though, I know I have done it before long ago).

Reply:
@gassmoeller I believe that the `while ... done < <(...)` really just follows the `command < file` structure that pipes the contents of a file into a command. The only part that is not commonly used is (i) that the command is here the `while ... done` loop, that `<(...)` isn't a real file but uses the output of another file -- see https://www.gnu.org/software/bash/manual/html_node/Process-Substitution.html#Process-Substitution . I had no idea that either works.

The problem with `ASTYLE_PATHS` in your example is if any of these paths contain spaces or, god forbid, `\0` characters which are apparently valid parts of file names. But I think that the `find ... | (while ... done)` syntax might be easier to read for sure, principally because it puts that which has to happen first (finding file names) *before* that which has to happen next (checking these file names).

Reply:
@gassmoeller @bangerth Problem with `find ... | while ...` is that for whatever reason it puts the loop block in its own background process, such that it cannot modify variables in the outer scope. Consider the following, for example:

    x='foo'
    find /bin -name 'bash' | while read exe; do x="$exe"; done
    echo "$x"
    >> foo

Apparently that can be avoided by first running `shopt -s lastpipe`, which instructs BASH to execute the loop contents within the current process, but I don't know whether that command is a standard part of BASH, or a linux specific thing. The process substitution version looks weird, but as Wolfgang noted, it's made of standard shell operations, and so should be as portable as possible.

Check out this stackoverflow thread for details: https://stackoverflow.com/a/23357277

On the other hand, I stand by my comment before that only a maniac would put whitespace characters in their PATH, so all of this seems pretty over the top anyway.

Reply:
Oh, then just do
```
  executable="`find ... | while ... echo $exe ... done`"
```

Reply:
Of course! I guess I got carried away. I'll update this after morning rounds.

Reply:
Someone in morning rounds mentioned that OSX apparently defaults to zsh now? The shebang in this file should force it to use bash if available, but while I was changing this, I added a commit (https://github.com/geodynamics/aspect/pull/3699/commits/5af4424b4f715944355c9ad31a8fad62f3676567) which makes this script also compatible with zsh (at least on linux). Tricky part was that zsh doesn't have an `export -f` equivalent, but if we're using the `find ... | while ...` construct here anyway, might as well use it for the dos2unix and tab replacement parts as well.

I can't believe I'm spending so much time generalizing something I could have fixed locally with a symlink! I'm gonna hold off on thinking about this until after the hackathon :)